### ISSUE TITLE
[Question] Issue in training VMAS simple spread with MAPPO and IPPO

### ISSUE THREAD

**Issue by kh-ryu** [facebookresearch/BenchMARL#144]
Hello, I'm trying to train a policy for VMAS simple spread environment using MAPPO and IPPO in Benchmarl. 
However, I'm suffering with some issues while training and it would be great if I can get any insight from you.

First, I tried MAPPO, using basic configuration given in BenchMARL. When I train, rewards and loss objectives are well behaved. 
![image](https://github.com/user-attachments/assets/5a9bcc37-801b-4e24-b8fd-aab0dd0e1ab6)
![image (1)](https://github.com/user-attachments/assets/39982d9f-c29c-40c5-9826-98c7738f8bd5)

However, at some point, it starts to output nan actions and I think the reason is because gradient of loss objective and clip friction goes beyond certain threshold that MAPPO can handle. 
![image (2)](https://github.com/user-attachments/assets/801a25cd-b07c-4c21-8425-90f35874f149)
![image (3)](https://github.com/user-attachments/assets/7c40adff-a11d-4de6-acc0-d614f99618dd)

On the other hand, when I train IPPO, mean reward starts to decrease at some point of the training
![image (4)](https://github.com/user-attachments/assets/862cb68e-94d1-41a4-be40-72592b81b993)

While I think this question belongs to general algorithm questions, I would appreciate if you can give any insight on why this happens and how to avoid this.

Thank you

**Comment by matteobettini** [comment]
Hello, thanks for opening this as I think this is a point that many users face and will benefit an answer to!

The short answer is that tuning RL algorithms is a bit like dark magic and especially with PPO there are these issues of catastrophic forgetting (the second one you mention) and gradients exploding (the first one you mention).

These issues do not normally pop up for me with the [vmas fine_tuned hyperparams](https://github.com/facebookresearch/BenchMARL/blob/main/fine_tuned/vmas/conf/config.yaml), but might occur in specific reward structures/ scenarios.

What normally happens with NaN actions is that the loss explodes, gradients explode, and weights go to nan in backpropagation, giving birth to nan actions. 

Here is some information on how to tackle this I gathered throughout the years (you are not the first user that encountered this, but I never made a public list):
- the entropy coefficient might be conflicting with optimality. maybe you are pushing for a higher entropy than what is required by the task?
- Have you tried using the vmas training parameters in the fine_tuned benchmarl folder? those work for me in the current vmas scenarios 
- highering the eps coefficient of adam https://github.com/facebookresearch/BenchMARL/blob/9813807e8ddd384e0a4dc2eecc411a208078de13/benchmarl/conf/experiment/base_experiment.yaml#L24 could help here with stability
- also watching the clip value for the gradient norm
- logging at what point the norm of your gradients start to explode could be helpful, maybe it is related to your agent entering a novel reward regime
- the scale of the MPE rewards is tuned quite badly, probably some reward tuning would help here (but I see that this might not make sense if you don't wnat to touch the original MPE task) 
- check if it happens also in `pettingzoo/simple_spread` task


Maybe try out some of these. but if the issue keeps happening with the default vmas/simple_spread and the default fine_tuned vmas params I can have a further look


Also, here are some points from @vmoens from the discord channel on the utilitiy of the loss metrics (maybe not relevant to your case but good to have)

> In RL the concept of a loss that always decreases isn't that relevant.
> Most of the time you don't have one but multiple losses (say one or more loss per model). These models are "competing" against each other. Let me take an example to make this clear:
> Imagine you're training your value network to predict the long-term return of your policy. 
> Initially, your reward is null or close to 0. The value network predicts a small return, and the L2 distance between that prediction and the true return is small. (similarly, if you're in a household with low income you can predict with a good accuracy that you will have low savings at the end of the year)
> Progressively, your policy becomes better, as a consequence the rewards have a higher magnitude, and so does the value that you're predicting. Now, the error of the value network is much bigger in magnitude, not because it's doing a worse job but because the landscape has changed. (taking the household example, if you're in a rich household the absolute value of the savings you can have at the end of the year has a much greater variance, not because you're worse at predicting it but because of the scale you're looking at).
> That means that the loss of your critic is actually getting worse through time (ie higher, not lower).
> 
> TL;DR: in RL the absolute value of the loss doesn't make sense. you will never see a paper reporting that. The only thing people care about is whether your reward/returns get better
> If I can add something here: in some cases (like PPO) the loss is even more pointless: in PPO the loss we compute is actually a proxy which gradient has the property of being on expectation equal to the gradient of some objective function that is intractable (an integral you can't solve analytically). In other words, for PPO the loss value is total garbage
> My fav metrics to look at in RL are:
> - reward/return curve: this is your P0 value, the only thing that really matters - Track this with RewardSum
> -  episode length (when it matters, eg CartPole) - Track this with StepCounter
> -  gradient norms of the various models (very high gradients or unexpectedly low may indicate ill behaved learning curves. Drastic changes can be worrying too if not correlated with better perf)
> -  exploding values in the loss values (per se the loss is meaningless but a value loss that goes to 10^38 probably means you're doing smth very wrong!)
>
> I general, any sudden change and/or extreme value requires an explanation
> 
>

**Comment by kh-ryu** [comment]
Hello @matteobettini 

Thank you for your kind answer! It helped me a lot on getting idea on hyperparameter tuning in RL.
As you suggested, I tried using fine_tuned hyperparameters and it fixed the problem.
Before that, I was using the parameter from [base_experiment.yaml](https://github.com/facebookresearch/BenchMARL/blob/main/benchmarl/conf/experiment/base_experiment.yaml).

It seems like main difference was minibatch_size, which was much bigger in fine_tuned hyperparam.

As I was able to fix my problem, I'll close this issue.

**Comment by matteobettini** [comment]
A point to add to this as I have now faced my issue myself in another context and none of the points above actually helped.

This seems to be usually related to the reward distribution. I such cases, normalizing the advantages in ppo or normalizing three rewards might actually help. Be careful tho that you normalize independently across agents.

See this for more info https://github.com/pytorch/rl/pull/2677

**Comment by zhaozilong2zl** [comment]
I encountered a similar return drop issue as @kh-ryu, and with the help of @matteobettini and the discussion in this issue, I finally managed to resolve my problem. Here’s my record:

## contex

I was using MAPPO in [custom simple_tag-like task](https://github.com/zhaozilong2zl/simple_tag_heaven_logic), where one side uses a scripted policy and the other side uses MAPPO, with a complex reward function composed of several components (including collision, pos_rew, velocity_direction_rew, etc.), aiming to train a multi-agent interception strategy.

## issue

I’ve encountered some strange phenomena, even when using the recommended [vmas fine_tuned hyperparams](https://github.com/facebookresearch/BenchMARL/blob/main/fine_tuned/vmas/conf/config.yaml)(see attached image and [wandb log](https://wandb.ai/2507844603/benchmarl/runs/mappo_simple_tag_heaven_complex_rew_negative_step_mlp__3f9a6294_25_03_05-15_35_14?nw=nwuser2507844603)). 

The episode_return_mean keeps increasing to a peak, but instead of stabilizing, it drops sharply almost immediately, eventually returning to a level similar to the starting point. I’ve tried adjusting the learning rate, lowering the task difficulty, tweaking reward function coefficients, and various other methods, but I always end up with this mountain-shaped curve. 

![Image](https://github.com/user-attachments/assets/b9a8c36a-dedb-43e7-b0a2-2cf066aa9fe3)

![Image](https://github.com/user-attachments/assets/acc63ce0-5f73-47ba-96bb-0c8cbd208ddd)


## one of the solution

You mentioned that reducing the minibatch_size resolved your sudden drop issue, which inspired me to think the problem might be related to data utilization. So, I tried reducing on_policy_n_minibatch_iters from 45 to 15, and it completely fixed my issue. the [return curve](https://wandb.ai/2507844603/benchmarl/runs/mappo_simple_tag_heaven_complex_rew_negative_step_mlp__2d1c84ce_25_03_06-13_06_26?nw=nwuser2507844603) i got no longer shows a drop (at least for now, though I haven’t tested if it might drop again with further training).

as @matteobettini pointed out:
> These issues do not normally pop up for me with the [vmas fine_tuned hyperparams](https://github.com/facebookresearch/BenchMARL/blob/main/fine_tuned/vmas/conf/config.yaml), but might occur in specific reward structures/ scenarios.

The biggest difference between my custom task and simple_tag is the reward function. I think this might be what caused the sudden drop, but I’m still not entirely sure about the exact reason. My approach is somewhat similar to @kh-ryu’s—reducing the batch_size in their case, while I reduced the reuse frequency. This might be more related to excessive learning, potentially leading to catastrophic forgetting or overfitting (if that’s the right way to describe it).
